kernel/src/main/scala/cats/kernel/instances/map.scala

```diff
@@ -5,13 +5,40 @@ import scala.collection.mutable
 
 package object map extends MapInstances
 
-trait MapInstances {
+trait MapInstances extends MapInstances1 {
+  implicit def catsKernelStdHashForMap[K: Hash, V: Hash]: Hash[Map[K, V]] =
+    new MapHash[K, V]
+}
+
+trait MapInstances1 {
   implicit def catsKernelStdEqForMap[K, V: Eq]: Eq[Map[K, V]] =
     new MapEq[K, V]
   implicit def catsKernelStdMonoidForMap[K, V: Semigroup]: Monoid[Map[K, V]] =
     new MapMonoid[K, V]
 }
 
+class MapHash[K, V](implicit V: Hash[V]) extends MapEq[K, V]()(V) with Hash[Map[K, V]] {
+  // adapted from [[scala.util.hashing.MurmurHash3]],
+  // but modified standard `Any#hashCode` to `ev.hash`.
+  import scala.util.hashing.MurmurHash3
+  def hash(x: Map[K, V]): Int = {
+    var a, b, n = 0
+    var c = 1
+    x foreach { case (k, v) =>
+      val h = (k.##, V hash v).##
```

Comment:
I don't like the `k.##` here. I guess this is because scala maps require universal hashing and equality, but it would be nice to add a comment about this.

Reply:
Yes, `k.##` is because Scala maps are using universal hashes. Also, I can optimize this tuple hashing to `Hash.caseClass2Hash("Tuple2".##, k.##, V.hash(v))`.